Optimize DB

amy opened the pull request:
https://github.com/rancher/rancher/issues/8716

Commit:
Do not display local duplicate catalog

Commit:
Add uuid to git.rancher.io repos for tracking

Commit:
git fetch optimization v1.6 compatible

Commit:
Merge pull request #93 from amy/v1.6

git fetch optimization v1.6 compatible

Commit:
Check in vendor code

Commit:
Fix Rancher API credentials

Adjust environment variables to match the injected environment variables
in https://github.com/rancher/cattle/pull/2801

Commit:
logging for catalog update

Commit:
Send UUID to mirror endpoint as a HTTP header

Commit:
Accept either http or https scheme and preserve it in formatted URL

Commit:
Remove redundant quotes

Commit:
revert to adding escaped quotes

Commit:
Revert Logging for catalog update

Commit:
Uuid tracking error handling and enhanced logging

- Die if cattle API is inaccessible, enhance logging
- Disable tracking/cattle connection conditionally
- Remove regexp that is no longer necessary

Commit:
Set track=false in test-warm

Commit:
Dont track when validating

Commit:
fix for airgap network catalog service panic issue

Commit:
close http response body

Commit:
Log and continue if install uuid can't be retrieved

Commit:
update v2 client

Commit:
vendor update

Commit:
Merge pull request #116 from cjellick/cherry-pick-clientUpdate

update to v2 client

cjellick commented on manager/db.go:
```diff
@@ -1,6 +1,12 @@
-import "github.com/rancher/catalog-service/model"
+import (
+	"fmt"
+
```
No need for this blank line. Impors from the go standard lib should be in a single group

cjellick commented on git/git.go:
```diff
@@ -87,6 +87,16 @@ func RemoteShaChanged(url, branch, sha, uuid string) bool {
+	pathSegments := strings.Split(path, "/")
+	gitRoot := fmt.Sprintf("/var/lib/cattle/%s", strings.Join(pathSegments[0:3], "/"))
+
+	cmd := exec.Command("git", fmt.Sprintf("--git-dir=%s/.git", gitRoot), "-C", gitRoot, "log", "-1", "--oneline", "--format=format:%H", "--", fmt.Sprintf("/var/lib/cattle/%s", path))
```
why do you need to set the git-dir here?

amy replied:
This is needed to run the git command in that respective git directory specified in the `--git-dir` flag

cjellick replied:
Why do none of our other git commands require it?

amy replied:
This git command needs to run in the respective git directory to get the history for changes in that directory.

cjellick commented:
We should consider profiling the improvement to DBs hit as a result of this PR

cjellick commented on manager/db.go:
```diff
@@ -60,128 +66,249 @@ func (m *Manager) lookupCatalog(environmentId, name string) (model.Catalog, erro
+	if reflect.DeepEqual(dbCatalog, model.CatalogModel{}) {
+		if err := tx.Create(&catalogModel).Error; err != nil {
+			tx.Rollback()
+			err := fmt.Errorf("CREATE CATALOG: %v %v %v", err.Error(), catalogModel.ID, dbCatalog)
```
this is a good opportunity to switch to `github.com/pkg/errors.Wrap()`

cjellick replied:
Is there a reason you formatted the error message this way with ACTION TABLE?

cjellick commented on manager/db.go:
```diff
@@ -60,128 +66,249 @@ func (m *Manager) lookupCatalog(environmentId, name string) (model.Catalog, erro
-		templateModel := model.TemplateModel{
-			Template: template,
+	// not found in db, create catalog
+	if reflect.DeepEqual(dbCatalog, model.CatalogModel{}) {
```
This code isn't very readable or self-evident. Is there a simpler, more obvious test you could do like `dbCatalog.ID == 0` that would more obviously indicate you are just checking to see if no result was found?

Doing so would also be more efficient than a reflect.DeepEqual (though that is admittedly a micro-optimization that would have negligible real impact).

amy replied:
I could write a separate function that will check if each field is a 0 value. My main concern with just checking if one of the values is the field's zero value is if that was intentional.

cjellick replied:
Can an ID field _be_ intentionally 0? it maps to a DB column, right?

cjellick commented on manager/db.go:
```diff
@@ -60,128 +66,249 @@ func (m *Manager) lookupCatalog(environmentId, name string) (model.Catalog, erro
 		}
-		if err := tx.Create(&templateModel).Error; err != nil {
+
+		tx.Where(&model.CatalogModel{
```
could you just set `dbCatalog = catalogModel` and save a DB lookup?

amy replied:
No. The main thing I need is the database catalog ID of the newly created catalog.

cjellick commented on manager/db.go:
```diff
@@ -60,128 +66,249 @@ func (m *Manager) lookupCatalog(environmentId, name string) (model.Catalog, erro
+
+	} else {
+		catalogModel.ID = dbCatalog.ID
+		if err := tx.Model(&model.CatalogModel{}).Where(&model.CatalogModel{
```
Why update based on name and envId as opposed to catalogModel.ID?

cjellick commented on manager/db.go:
```diff
@@ -60,128 +66,249 @@ func (m *Manager) lookupCatalog(environmentId, name string) (model.Catalog, erro
-					Category: model.Category{
-						Name: category,
+			// if template changes, change labels and categories
+			for k, v := range template.Labels {
```
Do we delete labels from the DB that the new template no longer has?

cjellick replied:
@amy this is my main concern/question

cjellick commented on manager/db.go:
```diff
@@ -60,128 +66,249 @@ func (m *Manager) lookupCatalog(environmentId, name string) (model.Catalog, erro
 }
 
 func (m *Manager) updateDb(catalog model.Catalog, templates []model.Template, newCommit string) error {
```
This function is huge. It would help the readability if you extract pieces of functionality into their own functions so that you could look at updateDb and get an idea of the overall logic and flow. This is over-simplifying and taking it to the extreme but it might look something like this:
```
udpateDb() {
  dbTemplates, dbTemplateCommits, dbTemplateLookups := getCatalogTemplates()
  createNewTemplates()
  deleteTemplates()
  syncLables()
  syncCategories()
}
```

amy replied:
This makes sense.

cjellick commented on manager/db.go:
```diff
@@ -60,128 +66,249 @@ func (m *Manager) lookupCatalog(environmentId, name string) (model.Catalog, erro
 					return err
 				}
+
+				for k, v := range version.Labels {
```
same label question: do you delete removed labels?

amy replied:
For version labels, they have version_id as a foreign key and have a cascaded delete. The cascade delete removes version labels and files.
 
Versions are deleted anytime there is a change.

Commit:
Update to go 1.8.3

This pulls in a change to automatically copy most headers on http
redirects. We had a custom redirect handler doing this, but can now
remove it.

Commit:
Refactor error handling

Swallow fewer errors.
Add meaningful context to errors.
Properly return 500 response codes on calls to the refresh API if the refresh fails.

Commit:
Fix test to not rely on order of catalogs

Commit:
Merge pull request #122 from cjellick/cherry-pick-error-handling

Cherry pick error handling

Commit:
db optimization

cjellick commented:
Very stale. Closing. Thanks! Please open a new PR if there is still a relevant change.
